### DISCUSSION TITLE
Understanding use_control_plane_lb = true

### DISCUSSION THREAD

**Discussion by clickersmudge** [mysticaltech/terraform-hcloud-kube-hetzner#302] [Q&A] [answered]
Hi,
I have three "control_plane_nodepools" and three "agent_nodepools".
When I enable "use_control_plane_lb = true" in the Hetnzer panel, two load balancers are created "test-control-plane" and "test".

Load Balancer "test-control-plane"
Target:
- test-control-plane-cx21-hel1-hmy
- test-control-plane-cx21-fsn1-tmx
- test-control-plane-cx21-nbg1-pih

Load Balancer "test"
- test-agent-cpx11-fsn1-xqd
- test-agent-cpx11-hel1-ekd
- test-storage-cpx21-fsn1-zgr
- test-agent-cpx11-nbg1-vvf

Shouldn't one "test-control-plane" Load Balancer be created that routes traffic between three "test-control-plane" and "test-control-plane that redirect traffic to  "test-agent"?
For which Load Balancer pinned a domain A record ("test-control-plane" or "test")?
When I check IP addresses
https://IP-test-control-plane (Doesn't open anything)
https://IP-test (Traefik 404 page open)
It seems to me that the "test-control-plane" should act as the main one but "test" throws a 404 page.
 I do not understand why?

**Comment by mysticaltech** [comment, marked as answer]
@clickersmudge Thanks for explaining what happens. By three control-plane-nodepools, I guess you mean three control-plane nodepools with one node in each. If that is so, the behavior is correct.

The only reason for activating `use_control_plane_lb` is to point your kubectl or any services that need to talk to the Kube API in an HA fashion. So it definitely will not respond to regular HTTP traffic. So really not needed in most cases. In the worst case, if the IP of one control-plane node is not responding for some reason, you can just hit the next one. Having an LB randomly point to a healthy control-plane node is a luxury unless you have mission-critical services needing that kind of HA to access the Kube API.

As for the regular lb, it will work great for you when you write an ingress definition and point it to one of your services that expose a deployment, for instance.

**Reply by clickersmudge** [reply]
Hi,
First of all, thank you for your time and detailed explanation.

Our setup is:
```
control_plane_nodepools = [
    {
      name        = "control-plane-cx21-fsn1",
      server_type = "cx21",
      location    = "fsn1",
      labels      = [],
      taints      = [],
      count       = 1
    },
    {
      name        = "control-plane-cx21-nbg1",
      server_type = "cx21",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 1
    },
    {
      name        = "control-plane-cpx21-hel1",
      server_type = "cpx21",
      location    = "hel1",
      labels      = [],
      taints      = [],
      count       = 1
    }
  ]
```
```
 agent_nodepools = [
    {
      name        = "agent-cpx11-fsn1",
      server_type = "cpx11",
      location    = "fsn1",
      labels      = [],
      taints      = [],
      count       = 1
    },
    {
      name        = "agent-cpx11-nbg1",
      server_type = "cpx11",
      location    = "nbg1",
      labels      = [],
      taints      = [],
      count       = 1
    },
    {
      name        = "agent-cpx11-hel1",
      server_type = "cpx11",
      location    = "hel1",
      labels      = [],
      taints      = [],
      count       = 1
    },
```

We want to achieve a configuration where Load Balancer will redirect us to one of the "control_plane_nodepools" in case one or two of them fails.

**Scenario "use_control_plane_lb = false"**
One Load Balancer is created in Hetnzer with "agent_nodepools" pinned to it.
Why then create three "control_plane_nodepools" for high availability?

**Scenario "use_control_plane_lb = true"**
Two Load Balancer is created in Hetnzer.
The first checks that "control_plane_nodepools" are working. 
The second one checks if "agent_nodepools" is working?
They are all in different regions. The best configuration for Super HA.
Am I right?

**Reply by mysticaltech** [reply]
@clickersmudge Your scenarios are correct!

Scenario "use_control_plane_lb = false" is still HA from a perspective of any services or application running on the cluster, as the cluster itself will never go down (as you have 3 CP nodes and +2 agents) as there will always be at least two CP nodes running at any time, as the node upgrades have a concurrency of 1, so only one can reboot at the same time. 

The only thing that can happen is that if you happen to hit the cluster with kubectl, at the moment when the first node is down for a reboot, for instance, kubectl will not work, so you have to manually edit the `kubeconfig.yaml` to point it to the next CP node.

As for the second scenario, it becomes HA even for cluster admins, as if one CP node is down for a reboot; for instance, kubectl, or any services relying on the Kube API (if you have those) will also work on the first try.

Now, it's important to remember that CP nodes are not meant to run workloads, but you can still do so by setting `allow_scheduling_on_control_plane` to true, which will allow the "normal LB" on scenario one to also point to CP nodes. But that's a different use case from scenario two, which focuses on making the Kube API HA, in other words, making the management of the cluster highly available.

Hopefully, that clarifies it!

**Reply by clickersmudge** [reply]
Thank you. This repository and yours work is brilliant.

**Reply by mysticaltech** [reply]
Thanks, @clickersmudge; I really appreciate those kind words! We tried out best 🙏

**Reply by maggie44** [reply]
Just a huge amount of information here, there are such rich discussions on this repo, it is helping move things along really quickly. 

A quick follow up on the HA of the control planes:

> The only reason for activating use_control_plane_lb is to point your kubectl or any services that need to talk to the Kube API in an HA fashion.

I assume this is applicable for HA when accessing the control planes externally. I am using the Kube API from inside a pod, would that already be HA?

**Reply by mysticaltech** [reply]
@maggie44 Good to hear! To answer the first part of your question, yes, the control plane LB is useful for external access. From inside the pod, at least as long as the pod is running, you should have access to the Kube API at all times. 

Now HA or not, it depends if you have at least 3 control-plane nodes or not.

**Reply by maggie44** [reply]
Great, thanks! I have 3 control-planes so should be all set 👍
